### PR TITLE
Adds support for different typed adjusted arrays and adds an EarningsCalendar loader

### PR DESCRIPTION
**Pull request by llllllllll**
- Moves most of AdjustedArray back into Python. The window iterator is
  the only part that's performance-intensive.

- Adds a bootleg templating system for creating specialized versions of
  AdjustedArrayWindow for each concrete type we care about.

- Adds support for differently dtyped terms in pipeline. This allows us
  to use datetime64s which are needed in the EarningsCalendar.

- Adds EarningsCalendar dataset for the next and previous earnings
  announcements in pipeline.

- Adds in memory loader for EarningsCalendar.

- Adds blaze loader for EarningsCalendar.

### CODE REVIEW

File: zipline/pipeline/loaders/blaze/earnings.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    def load_adjusted_array(self, columns, dates, assets, mask):
+        expr = self._expr
+        filtered = expr[expr[TS_FIELD_NAME] <= dates[0]]
+        lower = odo(
+            bz.by(
+                filtered[SID_FIELD_NAME],
+                timestamp=filtered[TS_FIELD_NAME].max(),
+            ).timestamp.min(),
+            pd.Timestamp,
+            **self._odo_kwargs or {}
```
Comment (ssanderson) [inline zipline/pipeline/loaders/blaze/earnings.py]: I don't think we need this check here.  We default to an empty dict in the constructor.

File: zipline/pipeline/loaders/blaze/earnings.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+            # range. It must all be null anyways.
+            lower = dates[0]
+
+        raw = odo(
+            expr[
+                (expr[TS_FIELD_NAME] >= lower) &
+                (expr[TS_FIELD_NAME] <= dates[-1])
+            ],
+            pd.DataFrame,
+            **self._odo_kwargs or {}
```
Comment (ssanderson) [inline zipline/pipeline/loaders/blaze/earnings.py]: Same note as above.

File: zipline/pipeline/loaders/blaze/earnings.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+        sids = raw.loc[:, SID_FIELD_NAME]
+        raw.drop(
+            sids[~(sids.isin(assets) | sids.notnull())].index,
+            inplace=True
+        )
+
+        gb = raw.groupby(SID_FIELD_NAME)
+        if self._has_ts:
+            def mkseries(idx, raw_loc=raw.loc):
```
Comment (ssanderson) [inline zipline/pipeline/loaders/blaze/earnings.py]: Is this just caching in the locals for performance?

File: zipline/pipeline/loaders/blaze/earnings.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        if not istabular(dshape):
+            raise ValueError(
+                'expression dshape must be tabular, got: %s' % dshape,
+            )
+
+        expected_fields = self._expected_fields
+        self._has_ts = has_ts = TS_FIELD_NAME in dshape.measure.dict
+        if not has_ts:
+            # This field is optional.
+            expected_fields - {TS_FIELD_NAME}
```
Comment (ssanderson) [inline zipline/pipeline/loaders/blaze/earnings.py]: I don't think this does anything...

File: zipline/pipeline/loaders/blaze/earnings.py (hunk 3 above)
Comment (llllllllll) [inline zipline/pipeline/loaders/blaze/earnings.py]: yeah, the function is only consumed within this scope so I figure this is safe.

File: tests/pipeline/test_numerical_expression.py [hunk 5]
```diff
@@ -128,6 +143,64 @@ def test_validate_bad(self):
         with self.assertRaises(TypeError):
             (f > f) > f
 
+    def test_combine_datetimes(self):
+        with self.assertRaises(TypeError) as e:
+            self.d + self.d
+        message = e.exception.args[0]
+        expected = (
+            "Don't know how to compute datetime64[ns] + datetime64[ns].\n"
+            "Arithmetic operators are only supported on Factors of dtype "
```
Comment (llllllllll) [inline tests/pipeline/test_numerical_expression.py]: Should this say `addition` instead. Subtraction is supported, no?

Comment (llllllllll) [inline tests/pipeline/test_numerical_expression.py]: also, is there a test for `datetime - datetime :: timedelta` or is this not yet supported?

File: zipline/lib/_windowtemplate.pxi [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+        self.data = data
+        self.viewtype = viewtype
+        self.adjustments = adjustments
+        self.adjustment_indices = sorted(adjustments, reverse=True)
+        self.window_length = window_length
+        self.anchor = window_length + offset
+        self.max_anchor = data.shape[0]
+
+        if len(self.adjustment_indices) > 0:
```
Comment (llllllllll) [inline zipline/lib/_windowtemplate.pxi]: this is duplicated in the `__next__`. Can we pull this out into a `next_adj` func?

File: tests/pipeline/test_numerical_expression.py (hunk 5 above)
Comment (ssanderson) [inline tests/pipeline/test_numerical_expression.py]: It's not supported yet, mostly because I couldn't think of a real use-case for it.

Comment (ssanderson) [inline tests/pipeline/test_numerical_expression.py]: (The error message is currently an accurate reflection of the actual behavior.)

Comment (llllllllll) [inline tests/pipeline/test_numerical_expression.py]: would you want to do something like:

```python
mask = EarningsCalendar.next_earnings.latest() - today < timedelta(5)
```

Comment (ssanderson) [inline tests/pipeline/test_numerical_expression.py]: You probably don't want calendar date arithmetic there though.  I added BusinessDaysUntil{Next,Previous}Earnings for that use case.

Comment (ssanderson) [inline tests/pipeline/test_numerical_expression.py]: And I'm not sure yet how we want to do `today` as a symbolic object.

File: zipline/lib/adjustment.pyx [hunk 7]
```diff
@@ -1,12 +1,116 @@
...
+from pandas import isnull, Timestamp
+from numpy cimport float64_t, uint8_t, int64_t
+from numpy import datetime64, float64
 # Purely for readability. There aren't C-level declarations for these types.
 ctypedef object Int64Index_t
 ctypedef object DatetimeIndex_t
 ctypedef object Timestamp_t
 
+# Adjustment kinds.
+MULTIPLY = 0
```
Comment (llllllllll) [inline zipline/lib/adjustment.pyx]: Could we define these with `cpdef enum AdjustmentKind:` for clarity. We can then change the usages of `int adjustment_kind` to `AdjustmentKind adjustment_kind`. This compiles the same but ls more clear in the source.

File: zipline/lib/adjustment.pyx [hunk 8]
```diff
@@ -1,12 +1,116 @@
...
+                                  Timestamp_t end_date,
+                                  int asset_id,
+                                  int adjustment_kind,
+                                  object value):
+    """
+    Make an Adjustment object from date/asset labels into a labelled baseline
+    array.
+    """
+    cdef type type_ = choose_adjustment_type(adjustment_kind, value)
+    return type_.from_assets_and_dates(
```
Comment (llllllllll) [inline zipline/lib/adjustment.pyx]: do we have the same issue with kwargs here?

Comment (ssanderson) [inline zipline/lib/adjustment.pyx]: Nope.  Yay Cython!

Comment (llllllllll) [inline zipline/lib/adjustment.pyx]: wow

File: zipline/pipeline/factors/events.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+    See Also
+    --------
+    BusinessDaysSincePreviousEarnings
+    """
+    inputs = [EarningsCalendar.next_announcement]
+    window_length = 0
+    dtype = float64_dtype
+
+    def _compute(self, arrays, dates, assets, mask):
```
Comment (llllllllll) [inline zipline/pipeline/factors/events.py]: since this is the same func as prev but with the operatands flipped, do you think it makes sense to share the setup code path? They are pretty short so idk if it is worth.

File: zipline/pipeline/factors/factor.py [hunk 10]
```diff
@@ -33,17 +28,67 @@
...
 
+def numbers_to_float64(func, argname, argvalue):
+    """
+    Preprocessor for converting numerical inputs into floats.
+
+    This is used in the binary operator constructors for Factor so that
+    `2 + Factor()` has the same behavior as `2.0 + Factor()`.
+    """
+    if isinstance(argvalue, Number):
+        return float64(argvalue)
```
Comment (llllllllll) [inline zipline/pipeline/factors/factor.py]: should we warn in the case of a possible loss of precisio? For example `argvalue=Fraction(1, 3)`

File: zipline/pipeline/term.py [hunk 11]
```diff
@@ -231,7 +274,8 @@ def _compute(self, windows, dates, assets, mask):
         """
         # TODO: Make mask available to user's `compute`.
         compute = self.compute
-        out = full(mask.shape, nan, dtype=self.dtype)
+        missing_value = self.missing_value
+        out = full(mask.shape, missing_value, dtype=self.dtype)
```
Comment (llllllllll) [inline zipline/pipeline/term.py]: `full_like(mask, ...)`

File: zipline/utils/input_validation.py [hunk 12]
```diff
@@ -30,6 +32,98 @@ def ensure_upper_case(func, argname, arg):
...
+    ...
+    >>> foo(arange(3), 'foo')
+    (array([0, 1, 2]), 'foo')
+    >>> foo(arange(3, dtype=float), 'foo')
+    Traceback (most recent call last):
+       ...
+    TypeError: foo() expected an argument with dtype 'int64' for argument 'x', but got dtype 'float64' instead.  # noqa
+    """
+    if _pos:
+        raise TypeError("expect_dtypes() only takes keyword arguments.")
```
Comment (llllllllll) [inline zipline/utils/input_validation.py]: why do we accept *args just to raise? This will raise a type error without accepting `_pos`.

File: zipline/utils/input_validation.py [hunk 13]
```diff
@@ -30,6 +32,98 @@ def ensure_upper_case(func, argname, arg):
...
+        "for argument '%(argname)s', but got %(actual)r instead."
+    ).format(dtype_str=' or '.join(repr(d.name) for d in allowed_dtypes))
+
+    def check_dtype(value):
+        return getattr(value, 'dtype', None) not in allowed_dtypes
+
+    def display_bad_value(value):
+        # If the bad value has a dtype, but it's wrong, show the dtype name.
+        if hasattr(value, 'dtype'):
+            return value.dtype.name
```
Comment (llllllllll) [inline zipline/utils/input_validation.py]: there are some cases where this will raise an attribute error, maybe check `hasattr(value, 'dtype') and hasattr(value.dtype, 'name')`

File: zipline/utils/numpy_utils.py [hunk 14]
```diff
@@ -1,7 +1,40 @@
...
+    float64_dtype: nan,
+    datetime64ns_dtype: np_NaT,
+}
+
+
+def default_fillvalue_for_dtype(dtype):
+    """
+    Get the default fill value for `dtype`.
+    """
+    return _FILLVALUE_DEFAULTS[dtype]
```
Comment (llllllllll) [inline zipline/utils/numpy_utils.py]: why don't consumers just use this dict directly?

File: zipline/utils/test_utils.py [hunk 15]
```diff
@@ -603,3 +605,28 @@ def wrapped(*args, **kwargs):
 
         return wrapped
     return dec
+
+
+def assert_timestamp_equal(left, right, compare_nat_equal=True, msg=""):
```
Comment (llllllllll) [inline zipline/utils/test_utils.py]: this is no longer used

File: zipline/pipeline/factors/events.py (hunk 9 above)
Comment (ssanderson) [inline zipline/pipeline/factors/events.py]: I thought about that while writing this and ended up just pulling out the functionality that's in `busday_count_mask_NaT`.  I kind of want to use these as examples of how to write factors that work with dates, so I think it's ok if they're a bit verbose/explicit.

File: zipline/pipeline/factors/factor.py (hunk 10 above)
Comment (ssanderson) [inline zipline/pipeline/factors/factor.py]: If numpy doesn't do this, then I'm not sure we should be in the business of doing it either.

Comment (ssanderson) [inline zipline/pipeline/factors/factor.py]:
```
In [2]: float64(Fraction(1, 3))
Out[2]: 0.33333333333333331
```

Comment (llllllllll) [inline zipline/pipeline/factors/factor.py]: Okay, just a thought. Also, numpy should totally warn about this.

**Comment by llllllllll** [comment]
other than the small comments your changes look good to merge.

File: zipline/utils/numpy_utils.py (hunk 14 above)
Comment (ssanderson) [inline zipline/utils/numpy_utils.py]: I can imagine the implementation becoming more complex in the future.  The fact that it's just a simple dict lookup right now is an implementation detail.

File: zipline/utils/input_validation.py (hunk 12 above)
Comment (ssanderson) [inline zipline/utils/input_validation.py]: This is how the other `expect_*` functions work as well.  I did this to provide a clearer error message I think, since I kept writing them passing positionally and getting confused when it didn't work.

I could see the argument for removing the *pos, but I don't think that that change belongs here.

Comment (llllllllll) [inline zipline/utils/input_validation.py]: Okay, I just think that it is not generally good practice to alias standard language errors. We can make this change is a different pr.

**Comment by ssanderson** [comment]
This is ready to merge here.  Waiting on downstream compat changes.
